### PR TITLE
Strict check advertise address

### PR DESCRIPTION
**Pull request by overvenus**
Address comments in #1335 .

PTAL @siddontang @BusyJay

### CODE REVIEW

**Comment by overvenus** [comment]
https://github.com/pingcap/tikv/pull/1345/commits/be30fe91d5c648aaab6f996aade4bbfef1563745 should be enough to fix #1340 .

File: src/bin/tikv-server.rs [hunk 1]
```diff
@@ -180,27 +180,20 @@ fn check_system_config(config: &toml::Value) {
...
         .unwrap();
 
-    let adv_addr = get_flag_string(matches, "advertise-addr")
-        .unwrap_or_else(|| get_toml_string(config, "server.advertise-addr", Some(addr.clone())));
-    util::config::check_addr(&adv_addr)
-        .map_err(|e| exit_with_err(format!("{:?}", e)))
-        .unwrap();
+    // FIXME: Forbiden addresses ending in 0 or 255? Those are not always invalid.
+    // See more: https://en.wikipedia.org/wiki/IPv4#Addresses_ending_in_0_or_255
+    let invalid_patterns = [("0.0.0.0", "0.0.0.0 is not allowed")];
```
Comment (siddontang) [inline src/bin/tikv-server.rs]: Do we need reason here? maybe following "invalid advertise-addr {}" is enough?

**Comment by siddontang** [comment]
Rest LGTM

File: src/bin/tikv-server.rs [hunk 2]
```diff
@@ -180,27 +180,20 @@ fn check_system_config(config: &toml::Value) {
...
+    util::config::check_addr(addr)
         .map_err(|e| exit_with_err(format!("{:?}", e)))
         .unwrap();
 
-    let adv_addr = get_flag_string(matches, "advertise-addr")
-        .unwrap_or_else(|| get_toml_string(config, "server.advertise-addr", Some(addr.clone())));
-    util::config::check_addr(&adv_addr)
-        .map_err(|e| exit_with_err(format!("{:?}", e)))
-        .unwrap();
+    // FIXME: Forbidden addresses ending in 0 or 255? Those are not always invalid.
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: When will these addresses become valid advertise addresses?

File: src/bin/tikv-server.rs [hunk 3]
```diff
@@ -180,27 +180,20 @@ fn check_system_config(config: &toml::Value) {
...
+    // See more: https://en.wikipedia.org/wiki/IPv4#Addresses_ending_in_0_or_255
+    let invalid_patterns = ["0.0.0.0"];
 
-    if let Some(_) = adv_addr.find("0.0.0.0") {
-        exit_with_err("0.0.0.0 is not allowed in advertise-addr".to_owned());
+    for pat in &invalid_patterns {
+        if let Some(_) = addr.find(pat) {
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: What if I pass in a "10.0.0.01"?

File: src/bin/tikv-server.rs [hunk 4]
```diff
@@ -180,27 +180,22 @@ fn check_system_config(config: &toml::Value) {
...
-    util::config::check_addr(&adv_addr)
-        .map_err(|e| exit_with_err(format!("{:?}", e)))
-        .unwrap();
+    // FIXME: Forbidden addresses ending in 0 or 255? Those are not always invalid.
+    // See more: https://en.wikipedia.org/wiki/IPv4#Addresses_ending_in_0_or_255
+    let invalid_patterns = [("0.", 0)]; // Current network is not allowed.
 
-    if let Some(_) = adv_addr.find("0.0.0.0") {
-        exit_with_err("0.0.0.0 is not allowed in advertise-addr".to_owned());
+    for &(pat, pos) in &invalid_patterns {
```
Comment (overvenus) [inline src/bin/tikv-server.rs]: I think `("0.", 0)` is okay, it is more flexible than `0.0.0.0`, plus `0.*.*.*` are definitely invalid.

File: src/bin/tikv-server.rs [hunk 5]
```diff
@@ -180,27 +180,22 @@ fn check_system_config(config: &toml::Value) {
...
 
-fn listen_address(matches: &Matches, config: &toml::Value) -> String {
-    let addr = get_flag_string(matches, "A").unwrap_or_else(|| {
-        get_toml_string(config,
-                        "server.addr",
-                        Some(DEFAULT_LISTENING_ADDR.to_owned()))
-    });
-    util::config::check_addr(&addr)
+fn check_advertise_address(addr: &str) {
+    util::config::check_addr(addr)
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: Use if let instead.

**Comment by ngaut** [comment]
PTAL @BusyJay @siddontang

**Comment by siddontang** [comment]
LGTM

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**